RIOT-OS/RIOT
# drivers/periph_gpio_ll: Fix GPIO_DISCONNECT handling and add compile time feature checks

### Contribution description

https://github.com/RIOT-OS/RIOT/pull/20189 enhanced the GPIO LL test to check if `GPIO_DISCONNECT` is also the state `gpio_ll_query_conf()` detects after applying this state using `gpio_ll_init()`. A number of GPIO LL implementation failed this tests, include STM32 (except for F1) and ATmega.

For ATmega the issue was that no dedicated "disconnected" state exists in hardware. Hence, `GPIO_DISCONNECT` was defined to have the same numeric value is `GPIO_INPUT`, which is the reset state for ATmega. This in turn resulted in compilation issues in code with `switch()` for the GPIO state in the common `gpio_ll_print_conf()` function. To address this, a compile time check was needed, which was done by introducing the feature `gpio_ll_disconnect`. While at it, the following features were all added:

 - `periph_gpio_ll_input_pull_down`:
        To indicate support for input mode with internal pull down
 - `periph_gpio_ll_input_pull_keep`:
        To indicate support for input mode with internal resistor pulling towards current level
 - `periph_gpio_ll_input_pull_up`:
        To indicate support for input mode with internal pull up
 - `periph_gpio_ll_disconnect`:
        To indicate a GPIO can be disconnected
 - `periph_gpio_ll_open_drain`:
        To indicate support for open drain mode
 - `periph_gpio_ll_open_drain_pull_up`:
        To indicate support for open drain mode with internal pull up
 - `periph_gpio_ll_open_source`:
        To indicate support for open source mode
 - `periph_gpio_ll_open_source_pull_down`:
        To indicate support for open source mode with internal pull down

### Testing procedure

The test app has been extended to check if features supported match features claimed. It will (as before) just query feature support at run time even if the features are not claimed to be supported. Then, it will `expect()` the support detected at run time match what features are claimed to be supported at compile time.

Note: For some MCUs the test currently are failing in `master`, due to `GPIO_DISCONNECT` not being correctly implemented.

### Issues/PRs references

https://github.com/RIOT-OS/RIOT/pull/20189 improved the test app that "introduced" the test failures by properly checking
